core/arch/arm/sm/pm.c

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017 NXP
+ *
+ * Peng Fan <peng.fan@nxp.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <arm32.h>
+#include <console.h>
+#include <drivers/imx_uart.h>
+#include <io.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/generic_boot.h>
+#include <kernel/misc.h>
+#include <kernel/panic.h>
+#include <kernel/pm_stubs.h>
+#include <kernel/tlb_helpers.h>
+#include <kernel/tz_ssvce_pl310.h>
+#include <mm/core_mmu.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <stdint.h>
+#include <sm/optee_smc.h>
+#include <sm/psci.h>
+#include <sm/sm.h>
+#include <sm/pm.h>
+
+/*
+ * See cpu_suspend_save argument 3
+ * save_p = &suspend_sp[0];
+ * Each cpu needs 2 words to save ptr and virt_to_phys(ptr).
+ * We max support 4 cpus now,so need 4 * 2 * sizeof(uint32_t);
+ */
+uint32_t suspend_sp[4 * 2] __aligned(8);
+
+void cpu_suspend_init(void)
+{
+	/* Place holer for now */
+}
+
+void cpu_suspend_save(uint32_t *ptr, uint32_t save_sz __unused, uint32_t sp,
+		      uint32_t *save_p)
+{
+	union {
+		void (*fun)(void);
+		void *addr;
+	} u;
+
+	/* save_p points to suspend_sp */
+	*save_p++ = (uint32_t)((void *)ptr);
+	*save_p = virt_to_phys((void *)ptr);
+
+	u.fun = cpu_do_resume;
+	/*
+	 * The content will be passed to cpu_do_resume
+	 * as register r1 and sp
+	 */
+	*ptr++ = 0; /* Not used now */
```

Comment:
It's not entirely clear to me.
We use structs at several places in the code that are accessed from assembly code, both for convenience and as documentation. There's `core/arch/arm/kernel/asm-defines.c` to help if needed.